### DISCUSSION TITLE
Hierarchical state machine: how to specify sub-state target?

### DISCUSSION THREAD

**Discussion by JakubKoralewski** [statelyai/xstate#2387] [Q&A] [answered]
Hi. I'm learning XState and have an issue with creating a hierarchical state machine with nested machines.

I'm trying to build a machine that consists of other machines. The code is closed sourced, but as an example let's take a situation where I have multiple chess boards and want to play on multiple of them concurrently, but first one has to be picked. Then from each board I can move freely to other chess boards and make moves on either of them one at at ime. Each chess board is in and of itself a state machine.

Here's how I went about approaching this:

1. Attempt 1. My initial state would be `BOARD_NOT_CHOSEN` and then after the user sends a `CHOSEN` event with data about which one they chose one the state would move to `BOARD_CHOSEN`. But then how do I spawn other state machines?
2. Attempt 2. Before constructing the xstate state machine I create an object where every unique key maps to a chessboard machine. Then I make the `BOARD_CHOSEN` state a hierarchical state machine which has all the states that I constructed before, i.e. each state specific to a single board is now a child-state of the `BOARD_CHOSEN` state. Each of those states has an `invoke` parameter which has a `src` of some `chessMachine` which hypothetically consists of the moves a single chess board can do.

I'm stuck on this cause I don't know how to move from `BOARD_NOT_CHOSEN` to `BOARD_CHOSEN` state, cause the initial state of the `BOARD_CHOSEN`'s child states is not known. The `target` property of the `machine.states.BOARD_NOT_CHOSEN.on.CHOSEN.target` is only a string, but I'd need to be able to specify something like `BOARD_CHOSEN.${evt.data}`. Not sure how to do this.

**Comment by johtso** [comment, marked as answer]
In the initial state of your parent machine your could spawn a machine for each of the chessboards, and store them in the context:

```ts
states: {
  noBoardSelected: {
    entry: assign((ctx) => ({
      boards: myBoards.map((board, i) => spawn(boardMachine.withContext(board), `board-${i}`)),
    })),
  ...
```

**Reply by JakubKoralewski** [reply]
But then it's like a parallel state machine (I can modify multiple boards at a time) and I lose the information on which board I am, I think.

**Reply by johtso** [reply]
You can just store the index of the current board in the context, and send events to that board.

Or if you'd rather not spawn all your boards at once, you could `invoke` the board when the board is selected: https://xstate.js.org/docs/guides/communication.html#invoking-machines

**Reply by johtso** [reply]
If the question is how to send an event to a specific child machine, you just do something like this:

```ts
send({ type: 'RESIGN' }, { to: (ctx) => ctx.currentBoard })
```
